[IV-23-24] Objetivo 4: Tests

johnwaves opened the pull request:
* [x]  ¿Hay un camino claro que permita ir desde el código hasta la historia de
   usuario que desarrolla, pasando por el mensaje de commit y el issue
   correspondiente?
* [x]  ¿Se ha intentado cubrir con tests una parte de la lógica de negocio que se
   esté desarrollando?
* [x]  ¿Hay un producto mínimamente viable que describa lo que se va a entregar en
   este objetivo?
* [x]  ¿En este producto mínimamente viable se han priorizado unas clases,
   módulos o paquetes más fundamentales frente a otros, movidos a otro
   PMV/Hito?
* [x] ¿Estoy leyendo todo lo que marco o simplemente marco lo que se me pone por
delante?
"Aprender sin reflexionar es malgastar la energía." - Confucio.
* [x] ¿Se han documentado las elecciones de biblioteca de aserciones y del
   de *test runner*?
* [x] ¿Te has asegurado que lo que mencionas como biblioteca de aserciones y
test runners lo son realmente y tienen la misma funcionalidad? ¿O estás poniendo
`unittest` como *test runner* y `pytest` como biblioteca de aserciones?
* [x] ¿Esa documentación incluye criterios de aceptación y también criterios de
búsqueda para las diferentes opciones?
* [x]  ¿Se han seguido los principios F.I.R.S.T. en la creación del test y
   se ha documentado cómo se ha hecho?
* [x]  ¿Se describe claramente en el PMV los tests que hay que pasar para que se
   considere viable?
* [x] ¿Se han testeado también las excepciones?

johnwaves committed:
Update README.md

johnwaves commented:
Cierro el issue #9 antes de seguir, pues ya quedó documentada la justificación de la elección del lenguaje de programación para el proyecto.

closes #9

johnwaves committed:
Añade archivo para elección de herramientas

Este archivo tendrá como objetivo recoger los motivos por los cuales se elegirá una u otra herramienta.

johnwaves committed:
Mueve archivo

johnwaves committed:
Actualiza información closes #48

Se añade la información encontrada sobre las diferentes bibliotecas de aserciones, test runners y frameworks existentes. También se especifica las herramientas seleccionadas y se proporciona una justificación en base a los criterios establecidos.

johnwaves committed:
Mueve estructura de dirección

Se eliminan en la estructura Cliente los campos relacionados con la
dirección y se crea una estructura para la dirección en un nuevo archivo.
También se añade un método para crear nuevas direcciones. El objetivo
es mejorar la legibilidad del código y posibilitar la reutilización de código.

johnwaves committed:
Elimina campos innecesarios #49

Los campos relacionados con la dirección del cliente se mueven a un
nuevo tipo llamado Dirección. Por otra parte, en la estructura Cliente
no es necesario almacenar tiempo_estimado_minutos. Esta variable resulta
inútil tenerla aquí, pues es necesaria en una estructura donde se encuentre
cada pedido almacenado junto a su tiempo estimado de entrega.

johnwaves committed:
Corrige variables closes #49

johnwaves committed:
Añade coordenadas de la ubicación del almacén #50

johnwaves committed:
Añade método para crear un nuevo cliente #49

johnwaves committed:
Añade métodos para obtener dirección #50

johnwaves committed:
Añade formato para dirección #50

johnwaves committed:
Implementa función para crear pedido #51

Se crea un nuevo tipo de estructura, ItemPedido, que será el que
utilice Pedido para almacenar los datos de los pedidos nuevos.
A su vez, cada pedido estará asociado a un único cliente, por lo que
se añade un campo de tipo Cliente a la estructura Pedido. Un cliente
podrá crear un pedido con varios items y este se le asignará automáticamente.

johnwaves committed:
Modifica campos de Ruta #52

Crea nueva estructura de datos para almacenar el pedido y la distancia
a la que se encuentra un cliente del almacén. Este dato se utilizará más
adelante para calcular la ruta de reparto.

johnwaves committed:
Añade cálculo de distancias closes #50

La mejor opción entre las barajadas para poder calcular la distancia
entre el almacén y la dirección de entrega de cada cliente es mediante
el método de Haversine, el cual obtiene la distancia (en kilómetros)
entre dos puntos de una esfera mediante sus coordenadas. Para obtener
estas coordenadas se usa una API de geocodificación, en este caso,
OpenStreetMap (OSM). Se añade también una función auxiliar para realizar
la conversión de grados a radianes.

johnwaves committed:
Reemplaza por función existente #50

johnwaves committed:
Crea estructura para coordenadas #50

johnwaves committed:
Reorganiza constantes para mejorar código closes #50

johnwaves committed:
Añade método para ruta óptima de reparto #52

El algoritmo de ordenación que se implementa se basa en la distnacia
existente entre el almacén y la ubicación de cada cliente. Aplicándolo,
los pedidos se dividen en tres categorías (cortos, medios y largos),
priorizando los lejanos si sólo hay de este tipo. Si no, se procede a
intercalar pedidos para equilibrar la atención entre lejanos y cercanos.
De esta forma, cada dos pedidos lejanos se intenta repartir un pedido corto.
Si no hay pedidos cortos disponibles, se rparte un pedido de distancia
media. Ese patrón se acaba repitiendo hasta que quedan todos ordenados.

johnwaves committed:
Añade explicación algoritmo closes #52

Se crea un documento donde se explica en detalle la solución propuesta
al problema del reparto equitativo de pedidos para este proyecto.

johnwaves committed:
Crea función para devolver items de un pedido closes #51

Mediante esta función se puede obtener los items de un pedido desde
cualquier paquete.

johnwaves committed:
Elimina límite inútil de pedidos

johnwaves committed:
Modifica funciones #53

Se modifican algunas funciones para aumentar la cobertura de pruebas.
En concreto, se añaden verificaciones de posibles errores que puedan
surgir al añadir pedidos a una ruta. También se muestran los errores.

johnwaves committed:
Añade test unitarios #53

Se añaden test para todas las funciones asociadas a la ruta. Se sigue el
estándar F.I.R.S.T. para su creación. Además, se comprueban funciones
auxiliares empleadas para obtener datos en la estructura Ruta.
También se ha instalado Testify en local.

johnwaves committed:
Actualiza información closes #53

carlotiii30 commented on docs/herramientas_test.md:
```diff
@@ -0,0 +1,36 @@
+- tener la capacidad de  automatizar los test para mejorar la eficiencia y la consistencia del proceso de prueba.
+
+## Listado de herramientas para la elección
+Considerando estos criterios, se evalúan múltiples opciones.
```
Creo que más que criterios has definido requisitos.

johnwaves replied:
Hola y gracias por la respuesta rápida.

Así es. Leyéndolo de nuevo parece que lo enfoqué más como requisitos y no definí como tal un criterio sólido. Lo voy a organizar de otra forma. Gracias de nuevo 😄.

JJ replied:
Por favor, resolved todos los comentarios. En este caso, cambiando la línea se habría resuelto; si lo que has cambiado es algo en otro lugar, tienes que resolverlo explícitamente.

carlotiii30 reviewed (changes requested):
Buenas! Creo que has trabajado bastante, sin embargo, hay un problema con los requisitos. Te he dejado un comentario que espero pueda ayudarte.

marinajcs commented on docs/herramientas_test.md:
```diff
@@ -0,0 +1,36 @@
+# Herramienta para la elaboración y gestión de tests y pruebas
+
+## Criterios
+La elección de herramientas para las pruebas y los tests que se realizarán antes de entregar un PMV deben seguir la misma línea de los criterios establecidos para los gestores de dependencias y tareas. Estas herramientas, en concreto y ante todo, deben ofrecer compatibilidad con Go, permitir una integración sin problemas con las demás medios y ofrecer facilidad para su configuración.
```
Ampliando lo mencionado por @carlotiii30, la compatibilidad con el lenguaje y toolchain actual quizás debería ser un requisito. 
Un ejemplo de criterio de selección podría ser el mantenimiento, importante para que las aserciones y frameworks no queden obsoletos. Estos aspectos puede evaluarse con buscadores como Snyk Advisor, que permite evaluar la calidad de las herramientas.

johnwaves replied:
Hola y gracias por responder tan rápido.

Sí, efectivamente, y como le he comentado a @carlotiii30, lo he enfocado de otra forma. Gracias también por la mención del buscador. Voy a echarle un vistazo a todo y a reorganizar el documento de otra forma 😄.

marinajcs reviewed (changes requested):
El PR tiene buena pinta en general, dejo aquí unas pequeñas sugerencias que espero que te ayuden a completar el objetivo 😃

johnwaves committed:
Modifica criterios de selección #48

Se modifica el enfoque usado para el detalle de los criterios de selección y se incluye también una lista de requisitos a cumplir que deben tener las herramientas de test.

johnwaves commented:
Lo he revisado y modificado. Gracias de nuevo por los comentarios, @carlotiii30 y @marinajcs! 😁

carlotiii30 reviewed (approved):
Muchas gracias por la reestructuración, lo veo mucho mejor.

marinajcs reviewed (approved):
No hay de qué, y gracias por considerar nuestro feedback, lo veo listo para revisión 😄

johnwaves commented:
@JJ listo para revisión.

JJ commented on docs/herramientas_test.md:
```diff
@@ -0,0 +1,39 @@
+
+### Criterios de selección
+- **Mantenimiento y actualizaciones regulares**. Las herramientas deben recibir actualizaciones y mantenimiento regulares para garantizar que no quedan obsoletas.
+- **Rendimiento de aserciones y pruebas**. Deben permitir la creación de aserciones rápidas, eficientes y flexibles.
```
Si no has testeado explícitamente, o consultado en algún artículo, el rendimiento, no puedes incluirlo. Incluid por favor sólo lo que hayáis considerado explícitamente a la hora de elegir. En tu elección de "Testify" no he visto que hayas mencionado en ningún momento su rendimiento.

JJ commented on docs/herramientas_test.md:
```diff
@@ -0,0 +1,39 @@
+### Criterios de selección
+- **Mantenimiento y actualizaciones regulares**. Las herramientas deben recibir actualizaciones y mantenimiento regulares para garantizar que no quedan obsoletas.
+- **Rendimiento de aserciones y pruebas**. Deben permitir la creación de aserciones rápidas, eficientes y flexibles.
+- **Generación de informes**. Otra cuestión de vital importancia es la capacidad que deben tener para generar informes o mostrar información relevante sobre los resultados de los tests.
```
Todos los tests runners tienen que generar informes, si no no son test runners.

JJ commented on docs/herramientas_test.md:
```diff
@@ -0,0 +1,39 @@
+
+### Requisitos básicos
+- **Compatibilidad con el lenguaje Go y su toolchain actual**. Las herramientas deben integrarse sin problemas con el entorno de desarrollo Go existente.
+- **Facilidad de configuración**. Deben ser fáciles de configurar y utilizar.
```
"facilidad" no es algo que se pueda evaluar fácilmente. En todo caso, la mayoría de las bibliotecas no requieren más configuración que incluir la dependencia.

JJ commented on docs/herramientas_test.md:
```diff
@@ -0,0 +1,39 @@
+- **Mantenimiento y actualizaciones regulares**. Las herramientas deben recibir actualizaciones y mantenimiento regulares para garantizar que no quedan obsoletas.
+- **Rendimiento de aserciones y pruebas**. Deben permitir la creación de aserciones rápidas, eficientes y flexibles.
+- **Generación de informes**. Otra cuestión de vital importancia es la capacidad que deben tener para generar informes o mostrar información relevante sobre los resultados de los tests.
+- **Soporte para diversos tipos de pruebas**. Deben ser versátiles para soportar diferentes tipos de pruebas según sea necesario.
```
¿Qué tipos de pruebas necesitas? En todo caso, ¿ de qué nivel de biblioteca estás hablando? No son lo mismo las bibliotecas de aserciones que los test runners, ¿no?

johnwaves replied:
Con "tipos de pruebas" me refería a pruebas que no sean únicamente de comprobar si un campo está vacío.

Y no, no son lo mismo las bibliotecas de aserciones que los test runners. Voy a intentar modificarlo de tal forma que no se entienda así nuevamente, en caso de haber ocurrido.

JJ commented on pkg/ruta.go:
```diff
@@ -1,5 +1,124 @@
+	}
+
+	direccion := cliente.direccion_entrega
+	if direccion.calle_numero == "" || direccion.poblacion == "" || direccion.codigo_postal == 0 {
```
¿No sería mejor que detectaras estos errores a nivel de `Pedido`? Si no te aseguras la corrección del objeto antes de pasarlo, estás cargando un montón de funciones adicionales con un montón de corrección de errores.

JJ commented on pkg/ruta.go:
```diff
@@ -1,5 +1,124 @@
+
+	direccion := cliente.direccion_entrega
+	if direccion.calle_numero == "" || direccion.poblacion == "" || direccion.codigo_postal == 0 {
+		return errors.New("Los datos de la dirección postal están incompletos.")
```
Por otro lado, no te va a hacer falta hasta que llegues al objetivo 7, pero si el error no dice exactamente *cuál* es el dato que está incompleto, no es útil ni para depurar ni para el cliente.

JJ commented on pkg/ruta.go:
```diff
@@ -1,5 +1,124 @@
+
+	for _, item := range pedido.items {
+		if item.cantidad <= 0 {
+			return errors.New("Cada pieza debe tener una cantidad mayor a cero.")
```
Igual, no estás diciendo cuál pieza tiene ese problema y por otro lado no estás garantizando en la creación del objeto que sea correcto.

JJ commented on pkg/ruta.go:
```diff
@@ -1,5 +1,124 @@
+	for _, pedido := range r.pedidos {
+		if pedido.Distancia < distanciaMedia {
+			pedidosCortos = append(pedidosCortos, pedido)
+		} else if pedido.Distancia < distanciaMaxima*0.75 {
```
No es buena práctica introducir literales en el código. Se deben definir constantes con un nombre significativo.

JJ commented on pkg/ruta.go:
```diff
@@ -1,5 +1,124 @@
+			pedidosCortos = append(pedidosCortos, pedido)
+		} else if pedido.Distancia < distanciaMaxima*0.75 {
+			pedidosMedios = append(pedidosMedios, pedido)
+		} else {
```
¿Es posible que `switch` sea un poco más sencillo de entender que estos if encadenados?

johnwaves replied:
Sí, tal vez resulte menos complejo.

JJ reviewed (changes requested):
Ante todo, muchas gracias a @carlotiii30 y a @marinajcs por la revisión; convendría en todo caso que le dieras una pequeña vuelta. La lógica de negocio parece sólida (aunque la programación es posiblemente un poco compleja) pero hay una serie de temas que te agradecería que consideraras.

johnwaves committed:
Cambiar literal por constante

johnwaves committed:
Reemplaza estructura if por switch #52

johnwaves committed:
Añade verificaciones #51

Se añaden varias verificaciones para notificar los errores que puedan
existir al crear un cliente y un pedido antes de llamar al método
AddPedido de Ruta.

johnwaves committed:
Elimina verificaciones #52

Elimina las antiguas verificaciones. Estas ahora se encuentran junto a
sus respectivas estructuras.

johnwaves committed:
Adapta tests a nuevas funciones #53

Tras añadir en algunas de las funciones existentes las verificaciones
oportunas, se han adaptado los tests para estas nuevas características.

JJ commented:
Convendría que corrigieras esto a la mayor brevedad. Ánimo.

johnwaves committed:
Actualiza información sobre criterios #48

Esta modificación intenta detallar de forma más clara qué criterios se
deben seguir y los que se han tenido en cuenta para la elección de la
herramienta de testeo.

johnwaves commented:
> Convendría que corrigieras esto a la mayor brevedad. Ánimo.

Listo de nuevo para revisión.

JJ reviewed (approved):
:+1:
